analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.PatternMatching.cs

```diff
@@ -190,5 +190,77 @@ public void RecursivePattern_SetsNotNull_Nested()
         }
+
+        [TestMethod]
+        public void DeclarationPattern_SetsNotNull_NoPreviousConstraint()
+        {
+            const string code = @"
+if (arg is Exception value)
+{
+    Tag(""Value"", value);
+    Tag(""ArgNotNull"", arg);
+}
+Tag(""End"", arg);";
+            var validator = SETestContext.CreateCS(code, ", object arg").Validator;
+            validator.ValidateContainsOperation(OperationKind.DeclarationPattern);
+            validator.ValidateTag("Value", x => x.HasConstraint(ObjectConstraint.NotNull).Should().BeTrue());
+            validator.ValidateTag("ArgNotNull", x => x.Should().BeNull());  // ToDo: MMF-2563 should have NotNull instead
+            validator.TagValues("End").Should().HaveCount(2).And.OnlyContain(x => x == null);       // 2x because value has different states
+        }
+
+        [TestMethod]
+        public void DeclarationPattern_SetsNotNull_PreservePreviousConstraint()
+        {
+            const string code = @"
+if (arg is Exception value)
+{
+    Tag(""Value"", value);
+    Tag(""ArgNotNull"", arg);
+}
+Tag(""End"", arg);";
+            var setter = new PreProcessTestCheck(OperationKind.ParameterReference, x => x.SetSymbolConstraint(x.Operation.Instance.TrackedSymbol(), TestConstraint.First));
+            var validator = SETestContext.CreateCS(code, ", object arg", setter).Validator;
+            validator.ValidateContainsOperation(OperationKind.DeclarationPattern);
+            validator.ValidateTag("Value", x => x.HasConstraint(TestConstraint.First).Should().BeTrue());
+            validator.ValidateTag("Value", x => x.HasConstraint(ObjectConstraint.NotNull).Should().BeTrue());
+            validator.ValidateTag("ArgNotNull", x => x.HasConstraint(TestConstraint.First).Should().BeTrue());
+            validator.ValidateTag("ArgNotNull", x => x.HasConstraint(ObjectConstraint.NotNull).Should().BeFalse());     // ToDo: MMF-2563 should BeTrue() instead
+            validator.TagValues("End").Should().HaveCount(2).And.OnlyContain(x => x != null && x.HasConstraint(TestConstraint.First));  // 2x because value has different states
+        }
+
+        [TestMethod]
+        public void DeclarationPattern_Discard_DoesNotFail()
+        {
+            const string code = @"
+if (arg is Exception _)
+{
+    Tag(""Arg"", arg);
+}
+Tag(""End"", arg);";
+            var validator = SETestContext.CreateCS(code, ", object arg").Validator;
+            validator.ValidateContainsOperation(OperationKind.DeclarationPattern);
+            validator.ValidateTag("Arg", x => x.Should().BeNull());  // ToDo: MMF-2563 should have NotNull instead
+            validator.TagValues("End").Should().HaveCount(1).And.OnlyContain(x => x == null);
+        }
+
+        [TestMethod]
+        public void DeclarationPattern_Var_PreservePreviousConstraint_DoesNotSetNotNullConstraint()
+        {
+            const string code = @"
+if (arg is var value)
```

Comment:
We should also make sure, we don't propagate to positional patterns (the same way we don't for property patterns)
`if (arg is var (x, y))`
`if (arg is (var x, var y))`
with parameter ", (int x, int y) arg"
This should be identical to what is tested in RecursivePattern_SetsNotNull_Nested *"Constraint from source value should not be propagated to child property"*